### PR TITLE
add test_backup_block_deletion

### PR DESCRIPTION
**Pull request by meldafrawi**
Longhorn/Longhorn#1435

### CODE REVIEW

File: manager/integration/tests/test_basic.py [hunk 1]
```diff
@@ -640,7 +645,78 @@ def test_backup_block_deletion():  # NOQA
...
+    volume = common.wait_for_volume_healthy(client, volume_name)
+
+    data0 = {'pos': 0,
+             'len': 4 * Mi,
+             'content': common.generate_random_data(4 * Mi)}
+
+    bv0, backup0, _, _ = create_backup(client, volume_name, data0)
+
+    data1 = {'pos': 0,
+             'len': 2 * Mi,
```
[the remark this answers is not in the document]
Comment (meldafrawi) [reply manager/integration/tests/test_basic.py]: fixed

File: manager/integration/tests/common.py [hunk 2]
```diff
@@ -3555,6 +3561,71 @@ def nfs_delete_random_backup_block(client, core_api, volume_name):
     pass
 
 
+def minio_count_backup_block_files(client, core_api, volume_name):
+    backup_target_credential_setting = client.by_id_setting(
+            SETTING_BACKUP_TARGET_CREDENTIAL_SECRET)
+
+    secret_name = backup_target_credential_setting.value
+
+    minio_api = get_minio_api(client, core_api, secret_name)
```
[the remark this answers is not in the document]
Comment (meldafrawi) [reply manager/integration/tests/common.py:3599]: fixed

File: manager/integration/tests/test_basic.py [hunk 3]
```diff
@@ -640,7 +631,72 @@ def test_backup_block_deletion():  # NOQA
...
+    bv1, backup1, _, _ = create_backup(client, volume_name, data1)
+
+    data2 = {'pos': 0,
+             'len': 2 * Mi,
+             'content': common.generate_random_data(2 * Mi)}
+
+    bv2, backup2, _, _ = create_backup(client, volume_name, data2)
+
+    backup_blocks_count = \
+        minio_count_backup_block_files(client, core_api, volume_name)
```
[the remark this answers is not in the document]
Comment (meldafrawi) [reply manager/integration/tests/test_basic.py]: fixed

File: manager/integration/tests/common.py [hunk 4]
```diff
@@ -3555,6 +3561,71 @@ def nfs_delete_random_backup_block(client, core_api, volume_name):
...
+
+
+def minio_delete_dummy_in_progress_backup(client, core_api, volume_name):
+    backup_target_credential_setting = client.by_id_setting(
+            SETTING_BACKUP_TARGET_CREDENTIAL_SECRET)
+
+    secret_name = backup_target_credential_setting.value
+
+    minio_api = get_minio_api(client, core_api, secret_name)
+    bucket_name = "backupbucket"
```
[the remark this answers is not in the document]
Comment (meldafrawi) [reply manager/integration/tests/common.py]: fixed

File: manager/integration/tests/common.py [hunk 5]
```diff
@@ -3552,7 +3573,121 @@ def minio_delete_random_backup_block(client, core_api, volume_name):
...
+
+    backup_cfg_file = "/tmp/backup_backup-dummy.cfg"
+
+    with open(backup_cfg_file, 'w') as bkp_cfg_file:
+        json.dump(dummy_backup_cfg_data, bkp_cfg_file)
+
+    try:
+        with open(backup_cfg_file, 'rb') as bkp_cfg_file:
+            bkp_cfg_file_stat = os.stat(backup_cfg_file)
+            minio_api.put_object(bucket_name,
```
Comment (joshimoo) [inline manager/integration/tests/common.py:3651]: unless the put_object needs a file handle, you can use
[json.dumps](https://docs.python.org/3/library/json.html#json.dumps) to turn the python object into a json string directly.

File: manager/integration/tests/common.py [hunk 6]
```diff
@@ -3552,7 +3573,121 @@ def minio_delete_random_backup_block(client, core_api, volume_name):
...
+    if is_backupTarget_s3(backupstore):
+        return minio_delete_dummy_in_progress_backup(client,
+                                                     core_api,
+                                                     volume_name)
+
+    elif is_backupTarget_nfs(backupstore):
+        return nfs_delete_dummy_in_progress_backup()
+
+
+def nfs_delete_dummy_in_progress_backup():
```
Comment (joshimoo) [inline manager/integration/tests/common.py:3672]: Can you add # TODO: comments and create an issue to add nfs support in the future.

File: manager/integration/tests/test_basic.py [hunk 7]
```diff
@@ -640,7 +632,69 @@ def test_backup_block_deletion():  # NOQA
...
+
+    data0 = {'pos': 0,
+             'len': 2 * BACKUP_BLOCK_SIZE_IN_MB * Mi,
+             'content': common.generate_random_data(4 * Mi)}
+
+    bv0, backup0, _, _ = create_backup(client, volume_name, data0)
+
+    data1 = {'pos': 0,
+             'len': BACKUP_BLOCK_SIZE_IN_MB * Mi,
+             'content': common.generate_random_data(2 * Mi)}
```
Comment (joshimoo) [inline manager/integration/tests/test_basic.py:650]: Why did you store the Backup block size not as bytes?
Then this call could look like this `common.generate_random_data(1 * BACKUP_BLOCK_SIZE)`

File: manager/integration/tests/test_basic.py [hunk 8]
```diff
@@ -640,7 +632,69 @@ def test_backup_block_deletion():  # NOQA
...
+    backupstore_cleanup(client)
+
+    volume = create_and_check_volume(client, volume_name)
+    host_id = get_self_host_id()
+    volume = volume.attach(hostId=host_id)
+    volume = common.wait_for_volume_healthy(client, volume_name)
+
+    data0 = {'pos': 0,
+             'len': 2 * BACKUP_BLOCK_SIZE_IN_MB * Mi,
+             'content': common.generate_random_data(4 * Mi)}
```
Comment (joshimoo) [inline manager/integration/tests/test_basic.py:644]: Then this call could look like this `common.generate_random_data(2 * BACKUP_BLOCK_SIZE)` with the line above `len: 2 * BACKUP_BLOCK_SIZE`

File: manager/integration/tests/test_basic.py [hunk 9]
```diff
@@ -640,7 +632,69 @@ def test_backup_block_deletion():  # NOQA
...
+    backup_blocks_count = backupstore_count_backup_block_files(client,
+                                                               core_api,
+                                                               volume_name)
+    assert backup_blocks_count == 4
+
+    bvs = client.list_backupVolume()
+
+    for bv in bvs:
+        if bv['name'] == volume_name:
+            assert bv['dataStored'] == str(backup_blocks_count * 2 * Mi)
```
Comment (joshimoo) [inline manager/integration/tests/test_basic.py:669]: backup_blocks_count * BACKUP_BLOCKS_SIZE seems nicer?

File: manager/integration/tests/common.py [hunk 10]
```diff
@@ -3552,7 +3573,121 @@ def minio_delete_random_backup_block(client, core_api, volume_name):
 
 # TODO: implement nfs_delete_random_backup_block
 def nfs_delete_random_backup_block(client, core_api, volume_name):
```
Comment (joshimoo) [inline manager/integration/tests/common.py:3575]: Do we currently limit the backup target to S3? in the test calling code, if so then this is okay.

File: manager/integration/tests/common.py (hunk 5 above)
Comment (meldafrawi) [reply manager/integration/tests/common.py:3651]: I've used `put_object` according to [minio's docs](https://docs.min.io/docs/python-client-api-reference.html#put_object)

File: manager/integration/tests/common.py (hunk 6 above)
Comment (meldafrawi) [reply manager/integration/tests/common.py:3672]: fixed

File: manager/integration/tests/test_basic.py (hunk 9 above)
Comment (meldafrawi) [reply manager/integration/tests/test_basic.py]: fixed

File: manager/integration/tests/test_basic.py (hunk 7 above)
Comment (meldafrawi) [reply manager/integration/tests/test_basic.py]: fixed

File: manager/integration/tests/test_basic.py (hunk 8 above)
Comment (meldafrawi) [reply manager/integration/tests/test_basic.py]: fixed

File: manager/integration/tests/common.py [hunk 11]
```diff
@@ -3517,26 +3532,32 @@ def get_minio_api(client, core_api, minio_secret_name):
...
-    secret_name = backup_target_credential_setting.value
-
-    minio_api = get_minio_api(client, core_api, secret_name)
-
+def minio_get_volume_backup_prefix(volume_name):
     volume_name_sha512 = \
         hashlib.sha512(volume_name.encode('utf-8')).hexdigest()
 
-    bucket_name = "backupbucket"
     backup_prefix = "backupstore/backupstore/volumes"
```
Comment (joshimoo) [inline manager/integration/tests/common.py:3539]: I think you need to pass the client and use get_backupstore_bucket_name ?

File: manager/integration/tests/common.py [hunk 12]
```diff
@@ -3517,26 +3532,32 @@ def get_minio_api(client, core_api, minio_secret_name):
...
+        volume_name
+
+    return prefix
+
+
+def minio_delete_random_backup_block(client, core_api, volume_name):
+    backup_target_credential_setting = client.by_id_setting(
+            SETTING_BACKUP_TARGET_CREDENTIAL_SECRET)
+
+    secret_name = backup_target_credential_setting.value
```
Comment (joshimoo) [inline manager/integration/tests/common.py:3555]: consider adding a get_backupstore_secret(client) ==> minio-secret

File: manager/integration/tests/common.py (hunk 11 above)
Comment (joshimoo) [reply manager/integration/tests/common.py:3539]: Or the s3 url? with the /path?

Comment (meldafrawi) [reply manager/integration/tests/common.py:3539]: I've already added that: [here](https://github.com/meldafrawi/longhorn-tests/blob/bd3ff1c8bfa851e5395b30a1e41ff02ad93188b0/manager/integration/tests/common.py#L3559)

File: manager/integration/tests/common.py (hunk 12 above)
Comment (meldafrawi) [reply manager/integration/tests/common.py]: fixed

File: manager/integration/tests/common.py (hunk 11 above)
Comment (joshimoo) [reply manager/integration/tests/common.py:3539]: I meant in regard to the backup_prefix = backupstore/backupstore/volumes
isn't the first /backupstore part of the s3 url?

Comment (joshimoo) [reply manager/integration/tests/common.py:3539]: ie for the engine I switched the target to /s3test

Comment (meldafrawi) [reply manager/integration/tests/common.py]: fixed

**Pull request merged**